gen/generate.go

```diff
@@ -200,72 +202,83 @@ func (i thriftPackageImporter) Package(file string) (string, error) {
 }
 
 func mergeFiles(dest, src map[string][]byte) error {
-	var errors []error
+	var err error
 	for path, contents := range src {
-		if _, ok := dest[path]; ok {
-			errors = append(errors, fmt.Errorf("file generation conflict: "+
-				"multiple sources are trying to write to %q", path))
-		}
-		dest[path] = contents
+		err = multierr.Append(err, addFile(dest, path, contents))
 	}
-	return multierr.Combine(errors...)
+	return err
 }
 
-// generateModule returns a mapping from filename to file contents of files that
-// should be generated relative to o.OutputDir.
-func generateModule(m *compile.Module, i thriftPackageImporter, builder *generateServiceBuilder, o *Options) (map[string][]byte, error) {
+func addFile(dest map[string][]byte, path string, contents []byte) error {
+	if _, ok := dest[path]; ok {
+		return fmt.Errorf("file generation conflict: "+
+			"multiple sources are trying to write to %q", path)
+	}
+	dest[path] = contents
+	return nil
+}
+
+// generateModule generates the code for the given Thrift file and returns the
+// path to the output file relative to OutputDir and the contents of the file.
+func generateModule(
+	m *compile.Module,
+	i thriftPackageImporter,
+	builder *generateServiceBuilder,
+	o *Options,
+) (outputFilepath string, contents []byte, err error) {
 	// packageRelPath is the path relative to outputDir into which we'll be
 	// writing the package for this Thrift file. For $thriftRoot/foo/bar.thrift,
 	// packageRelPath is foo/bar, and packageDir is $outputDir/foo/bar. All
 	// files for bar.thrift will be written to the $outputDir/foo/bar/ tree. The
 	// package will be importable via $importPrefix/foo/bar.
 	packageRelPath, err := i.RelativePackage(m.ThriftPath)
 	if err != nil {
-		return nil, err
+		return "", nil, err
 	}
 
 	// TODO(abg): Prefer top-level package name from `namespace go` directive.
 	packageName := filepath.Base(packageRelPath)
 
+	// Output file name defaults to the package name.
+	outputFilename := packageName + ".go"
+	if len(o.OutputFile) > 0 {
+		outputFilename = o.OutputFile
+	}
+	outputFilepath = filepath.Join(packageRelPath, outputFilename)
+
 	// importPath is the full import path for the top-level package generated
 	// for this Thrift file.
 	importPath, err := i.Package(m.ThriftPath)
 	if err != nil {
-		return nil, err
+		return "", nil, err
 	}
 
-	// Mapping of file names relative to packageRelPath to their contents.
-	// Note that we need to return a mapping relative to o.OutputDir so we
-	// will prepend $packageRelPath/ to all these paths.
-	files := make(map[string][]byte)
-
-	gopts := &GeneratorOptions{
+	g := NewGenerator(&GeneratorOptions{
```

Comment:
Thank you for cleaning up the code! It looks much cleaner with the Options object declared in the NewGenerator called!